src/core/storage/sframe_interface/unity_sframe.cpp

```diff
@@ -1200,6 +1201,73 @@ unity_sframe::random_split(float percent, uint64_t random_seed, bool exact) {
   return logical_filter_split(logical_filter_array);
 }
 
+std::shared_ptr<unity_sframe_base> unity_sframe::shuffle() {
+#include <iostream>  // TODO: remove
+  
+  log_func_entry();
+
+  std::shared_ptr<unity_sframe> temp(new unity_sframe());
+
+  size_t length = this->size();
+
+  std::vector<std::string> column_names = this->column_names();
+
+  const size_t num_buckets = 2;   // XXX: how to determine the number of buckets?
+
+  // Create a column of random ints between 0 and (num_buckets - 1).
+  auto temp_groupby_column = std::static_pointer_cast<unity_sarray>(
+    unity_sarray::make_uniform_int_array(length, num_buckets));
+  const std::string rand_int_column_name("Random Ints");
+  temp->add_column(temp_groupby_column, rand_int_column_name);
+
+  // Pack columns so we can group by concatenate
+  std::shared_ptr<unity_sarray_base> packed_columns = this->pack_columns(column_names,
+                                                                         column_names,
+                                                                         flex_type_enum::LIST,
+                                                                         turi::flex_undefined());
+  const std::string packed_data_column_name("Packed Data");
+  temp->add_column(packed_columns, packed_data_column_name);
+
+  // Group by concatenate on the random int column. This randomly bucketizes.
+  const std::string buckets_column_name("Buckets");
+  std::shared_ptr<unity_sframe_base> bucketized_sframe = temp->groupby_aggregate({rand_int_column_name},
+                                                                                 {{packed_data_column_name}},
+                                                                                 {buckets_column_name},
+                                                                                 {"__builtin__concat__list__"});
+  std::shared_ptr<unity_sarray_base> bucketized_sarray = bucketized_sframe->select_column(buckets_column_name);
+
+  // XXX: do this in parallel
```

Comment:
Get it working first, then let's parallelize it.

Reply:
To do this in parallel, use this code template: 
```
size_t num_threads = thread::cpu_count();

gl_sarray_writer writer(flex_type_enum::LIST, num_threads);

in_parallel([&](size_t thread_id, size_t n_threads) { 
...
size_t idx_start = (gl_bucketized_sarray.num_rows() * thread_idx) / num_threads;
size_t idx_end = (gl_bucketized_sarray.num_rows() * (thread_idx + 1) ) / num_threads;
 
gl_sarray_range ra = gl_bucketized_sarray.range_iterator(idx_start, idx_end);

// Same code below, just write to the correct segment.

});

// Now close the writer
```

Reply:
Thanks @hoytak - I was able to use this code almost verbatim.